src/main/java/seedu/address/model/SerialNumberGenerator.java

```diff
@@ -11,44 +12,46 @@
     public static final int SERIAL_NUMBER_LENGTH = 5; //excluding prefix 'B'
     public static final String PREFIX = "B";
 
-    private static int currentSerialNumberIndex = 0;
-
-    private static Catalog catalog = new Catalog();
+    private static TreeSet<SerialNumber> serialNumberTree = new TreeSet<>();
 
+    /**
+     * Populates the serial number tree from a catalog.
+     *
+     * @param catalog catalog to retrieve books from.
+     */
     public static void setCatalog(Catalog catalog) {
-        SerialNumberGenerator.catalog = catalog;
-        currentSerialNumberIndex = 0;
+        serialNumberTree = new TreeSet<>();
+        catalog.getBookList()
+                .forEach(book -> serialNumberTree.add(book.getSerialNumber()));
     }
+
     /**
      * Generates a new serial number based on the current serial number index.
      */
     public static SerialNumber generateSerialNumber() {
-        currentSerialNumberIndex++;
-        int paddingLength = getPaddingLength();
-        String padding = getPadding(paddingLength);
-        SerialNumber sn = new SerialNumber(PREFIX + padding + currentSerialNumberIndex);
-        while (catalog.checkIfSerialNumberExists(sn)) {
-            currentSerialNumberIndex++;
-            sn = new SerialNumber(PREFIX + padding + currentSerialNumberIndex);
+        if (serialNumberTree.isEmpty()) {
+            SerialNumber serialNumber = new SerialNumber(PREFIX + "00001");
+            serialNumberTree.add(serialNumber);
+            return serialNumber;
         }
-        return new SerialNumber(PREFIX + padding + currentSerialNumberIndex);
+        int key = serialNumberTree.size();
+        SerialNumber keyCompare = constructSerialNumberFromInt(key);
+        int newIndex = serialNumberTree.floor(keyCompare).serialNumberToInt() + 1;
```

Comment:
Will there be a potential bug when floor returns null?
Let's say my current tree only has 3 items, which are serial numbers 8, 9, 10. (The other books are deleted or smthg.)
Since size = 3, calling floor(3) would return null.

Reply:
Good catch! Shall add an additional catch in this case